File: 00_Alternate_Languages/85_Synonym/java/src/Synonym.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

/**
 * Game of Synonym
 * <p>
 * Based on the Basic game of Synonym here
 * https://github.com/coding-horror/basic-computer-games/blob/main/85%20Synonym/synonym.bas
 * <p>
 * Note:  The idea was to create a version of the 1970's Basic game in Java, without introducing
 * new features - no additional text, error checking, etc has been added.
 */
public class Synonym {

    public static final String[] RANDOM_ANSWERS = {"RIGHT", "CORRECT", "FINE", "GOOD!", "CHECK"};

    // Used for keyboard input
    private final Scanner kbScanner;

    // List of words and synonyms
    private final ArrayList<SynonymList> synonyms;

    private enum GAME_STATE {
        INIT,
        PLAY,
        GAME_OVER
    }

    // Current game state
    private GAME_STATE gameState;

    private int currentQuestion;

    public Synonym() {

        kbScanner = new Scanner(System.in);
        synonyms = new ArrayList<>();

        gameState = GAME_STATE.INIT;
    }

    /**
     * Main game loop
     */
    public void play() {

        do {
            switch (gameState) {

                case INIT:
                    intro();
                    currentQuestion = 0;

                    // Load data
                    synonyms.add(new SynonymList("FIRST", new String[]{"START", "BEGINNING", "ONSET", "INITIAL"}));
                    synonyms.add(new SynonymList("SIMILAR", new String[]{"SAME", "LIKE", "RESEMBLING"}));
                    synonyms.add(new SynonymList("MODEL", new String[]{"PATTERN", "PROTOTYPE", "STANDARD", "CRITERION"}));
                    synonyms.add(new SynonymList("SMALL", new String[]{"INSIGNIFICANT", "LITTLE", "TINY", "MINUTE"}));
                    synonyms.add(new SynonymList("STOP", new String[]{"HALT", "STAY", "ARREST", "CHECK", "STANDSTILL"}));
                    synonyms.add(new SynonymList("HOUSE", new String[]{"DWELLING", "RESIDENCE", "DOMICILE", "LODGING", "HABITATION"}));
                    synonyms.add(new SynonymList("PIT", new String[]{"HOLE", "HOLLOW", "WELL", "GULF", "CHASM", "ABYSS"}));
                    synonyms.add(new SynonymList("PUSH", new String[]{"SHOVE", "THRUST", "PROD", "POKE", "BUTT", "PRESS"}));
                    synonyms.add(new SynonymList("RED", new String[]{"ROUGE", "SCARLET", "CRIMSON", "FLAME", "RUBY"}));
                    synonyms.add(new SynonymList("PAIN", new String[]{"SUFFERING", "HURT", "MISERY", "DISTRESS", "ACHE", "DISCOMFORT"}));

                    gameState = GAME_STATE.PLAY;
                    break;

                case PLAY:

                    // Get the word and synonyms to ask a question about
                    SynonymList synonym = synonyms.get(currentQuestion);
                    String getAnswer = displayTextAndGetInput("     WHAT IS A SYNONYM OF " + synonym.getWord() + " ? ");

                    // HELP is used to give a random synonym for the current word
                    if (getAnswer.equals("HELP")) {
                        int randomSynonym = (int) (Math.random() * synonym.size());
                        System.out.println("**** A SYNONYM OF " + synonym.getWord() + " IS " + synonym.getSynonyms()[randomSynonym] + ".");
                    } else {
                        // Check if the entered word is in the synonym list
                        if (synonym.exists(getAnswer)) {
                            // If it is, give a random "correct" response
                            System.out.println(RANDOM_ANSWERS[(int) (Math.random() * RANDOM_ANSWERS.length)]);
                            currentQuestion++;
                            // Have we reached the final word/synonyms on file?
                            if (currentQuestion == synonyms.size()) {
                                // We have so end game.
                                System.out.println("SYNONYM DRILL COMPLETED.");
                                gameState = GAME_STATE.GAME_OVER;
                            }
                        } else {
                            // Word does not exist in the synonym list
                            System.out.println("TRY AGAIN.");
                        }
                    }
            }
        } while (gameState != GAME_STATE.GAME_OVER);
    }

    private void intro() {
        System.out.println(simulateTabs(33) + "SYNONYM");
        System.out.println(simulateTabs(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY");
        System.out.println();
        System.out.println("A SYNONYM OF A WORD MEANS ANOTHER WORD IN THE ENGLISH");
        System.out.println("LANGUAGE WHICH HAS THE SAME OR VERY NEARLY THE SAME");
        System.out.println(" MEANING.");
        System.out.println("I CHOOSE A WORD -- YOU TYPE A SYNONYM.");
        System.out.println("IF YOU CAN'T THINK OF A SYNONYM, TYPE THE WORD 'HELP'");
        System.out.println("AND I WILL TELL YOU A SYNONYM.");
        System.out.println();
    }

    /*
     * Print a message on the screen, then accept input from Keyboard.
     * Converts input to uppercase.
     *
     * @param text message to be displayed on screen.
     * @return what was typed by the player.
     */
    private String displayTextAndGetInput(String text) {
        System.out.print(text);
        return kbScanner.next().toUpperCase();
    }

    /**
     * Simulate the old basic tab(xx) command which indented text by xx spaces.
     *
     * @param spaces number of spaces required
     * @return String with number of spaces
     */
    private String simulateTabs(int spaces) {
        char[] spacesTemp = new char[spaces];
        Arrays.fill(spacesTemp, ' ');
        return new String(spacesTemp);
    }
}

File: 00_Alternate_Languages/85_Synonym/javascript/synonym.js
// SYNONYM
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;

    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

var ra = [, "RIGHT", "CORRECT", "FINE", "GOOD!", "CHECK"];
var la = [];
var tried = [];

var synonym = [[5,"FIRST","START","BEGINNING","ONSET","INITIAL"],
               [5,"SIMILAR","ALIKE","SAME","LIKE","RESEMBLING"],
               [5,"MODEL","PATTERN","PROTOTYPE","STANDARD","CRITERION"],
               [5,"SMALL","INSIGNIFICANT","LITTLE","TINY","MINUTE"],
               [6,"STOP","HALT","STAY","ARREST","CHECK","STANDSTILL"],
               [6,"HOUSE","DWELLING","RESIDENCE","DOMICILE","LODGING","HABITATION"],
               [7,"PIT","HOLE","HOLLOW","WELL","GULF","CHASM","ABYSS"],
               [7,"PUSH","SHOVE","THRUST","PROD","POKE","BUTT","PRESS"],
               [6,"RED","ROUGE","SCARLET","CRIMSON","FLAME","RUBY"],
               [7,"PAIN","SUFFERING","HURT","MISERY","DISTRESS","ACHE","DISCOMFORT"]
               ];

// Main program
async function main()
{
    print(tab(33) + "SYNONYM\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    for (c = 0; c <= synonym.length; c++)
        tried[c] = false;
    print("A SYNONYM OF A WORD MEANS ANOTHER WORD IN THE ENGLISH\n");
    print("LANGUAGE WHICH HAS THE SAME OR VERY NEARLY THE SAME");
    print(" MEANING.\n");
    print("I CHOOSE A WORD -- YOU TYPE A SYNONYM.\n");
    print("IF YOU CAN'T THINK OF A SYNONYM, TYPE THE WORD 'HELP'\n");
    print("AND I WILL TELL YOU A SYNONYM.\n");
    print("\n");
    c = 0;
    while (c < synonym.length) {
        c++;
        do {
            n1 = Math.floor(Math.random() * synonym.length + 1);
        } while (tried[n1]) ;
        tried[n1] = true;
        n2 = synonym[n1][0];    // Length of synonym list
        // This array keeps a list of words not shown
        for (j = 1; j <= n2; j++)
            la[j] = j;
        la[0] = n2;
        g = 1;  // Always show first word
        print("\n");
        la[g] = la[la[0]];  // Replace first word with last word
        la[0] = n2 - 1; // Reduce size of list by one.
        print("\n");
        while (1) {
            print("     WHAT IS A SYNONYM OF " + synonym[n1][g]);
            str = await input();
            if (str == "HELP") {
                g1 = Math.floor(Math.random() * la[0] + 1);
                print("**** A SYNONYM OF " + synonym[n1][g] + " IS " + synonym[n1][la[g1]] + ".\n");
                print("\n");
                la[g1] = la[la[0]];
                la[0]--;
                continue;
            }
            for (k = 1; k <= n2; k++) {
                if (g == k)
                    continue;
                if (str == synonym[n1][k])
                    break;
            }
            if (k > n2) {
                print("     TRY AGAIN.\n");
            } else {
                print(synonym[n1][Math.floor(Math.random() * 5 + 1)] + "\n");
                break;
            }
        }
    }
    print("\n");
    print("SYNONYM DRILL COMPLETED.\n");
}

main();
